tests/src/core/testqgsmaprendererjob.cpp

```diff
@@ -1075,6 +1083,121 @@ void TestQgsMapRendererJob::customNullPainterJob()
+  vectorLayer->commitChanges();
+  QVERIFY( vectorLayer->featureCount() == 1 );
+  std::unique_ptr<QgsFillSymbol> fill( static_cast< QgsFillSymbol * >( QgsSymbol::defaultSymbol( QgsWkbTypes::PolygonGeometry ) ) ) ;
+  fill->setColor( QColor( 255, 255, 255 ) );
+  vectorLayer->setRenderer( new QgsSingleSymbolRenderer( fill.release() ) );
+
+  QgsMapSettings mapSettings;
+  mapSettings.setDestinationCrs( pointCloudLayer->crs() );
+  mapSettings.setExtent( pointCloudLayer->extent() );
+  mapSettings.setOutputSize( QSize( 512, 512 ) );
+  mapSettings.setOutputDpi( 96 );
+  mapSettings.setLayers( QList< QgsMapLayer * >()
+                         << vectorLayer.get()
+                         << pointCloudLayer.get()
+                         << rasterLayer.get()
+                         << meshLayer.get() );
+
+  QgsShadingRenderer shadingRenderer;
+  shadingRenderer.setActive( true );
+  shadingRenderer.setActiveHillShading( true );
+  shadingRenderer.setActiveEyeDomeLighting( false );
+  mapSettings.setShadingRenderer( shadingRenderer );
+  std::unique_ptr<QgsMapRendererSequentialJob> renderJob( new QgsMapRendererSequentialJob( mapSettings ) );
+  renderJob->start();
+  renderJob->waitForFinished();
+  QImage img = renderJob->renderedImage();
+  QVERIFY( imageCheck( QStringLiteral( "render_shading_1" ), img ) );
+
+  shadingRenderer.setLightAltitude( 20 );
+  shadingRenderer.setLightAzimuth( 60 );
+  mapSettings.setShadingRenderer( shadingRenderer );
+  renderJob.reset( new QgsMapRendererSequentialJob( mapSettings ) );
+  renderJob->start();
+  renderJob->waitForFinished();
+  img = renderJob->renderedImage();
+  QVERIFY( imageCheck( QStringLiteral( "render_shading_2" ), img ) );
+
+  shadingRenderer.setHillShadingMultidirectional( true );
+  shadingRenderer.setHillShadingZFactor( 5 );
+  mapSettings.setShadingRenderer( shadingRenderer );
+  renderJob.reset( new QgsMapRendererSequentialJob( mapSettings ) );
+  renderJob->start();
+  renderJob->waitForFinished();
+  img = renderJob->renderedImage();
+  QVERIFY( imageCheck( QStringLiteral( "render_shading_3" ), img ) );
+
+  shadingRenderer.setCombinedElevationMethod( QgsElevationMap::CombineMethod::NewerElevation );
+  shadingRenderer.setActiveHillShading( false );
+  shadingRenderer.setActiveEyeDomeLighting( true );
+  mapSettings.setShadingRenderer( shadingRenderer );
+  renderJob.reset( new QgsMapRendererSequentialJob( mapSettings ) );
+  renderJob->start();
+  renderJob->waitForFinished();
+  img = renderJob->renderedImage();
+  QVERIFY( imageCheck( QStringLiteral( "render_shading_4" ), img ) );
+
+  shadingRenderer.setEyeDomeLightingDistance( 10 );
+  shadingRenderer.setEyeDomeLightingStrength( 4000 );
+  mapSettings.setShadingRenderer( shadingRenderer );
+  renderJob.reset( new QgsMapRendererSequentialJob( mapSettings ) );
+  renderJob->start();
+  renderJob->waitForFinished();
+  img = renderJob->renderedImage();
+  QVERIFY( imageCheck( QStringLiteral( "render_shading_5" ), img ) );
+
```

Comment:
why are all the resulting rendered images grayscale only? I would expect they would be combined with color image from the "normal" map rendering - to make sure that this combination works correctly...

Reply:
ah... The idea was to have a vector layer on the top the see if all layer are shaded, and I set the symbology of this this layer white... Not such a good idea. I will set the vector layer between others and add a bit of colors.